source/slang/slang-check-conversion.cpp

```diff
@@ -1866,6 +1870,65 @@ bool SemanticsVisitor::_coerce(
     return _failedCoercion(toType, outToExpr, fromExpr, sink);
 }
 
+// TODO make member of SemanticsVisitor
+bool SemanticsVisitor::tryCoerceSomeType(
+    CoercionSite site,
+    Type* toType,
+    Expr** outToExpr,
+    QualType fromType,
+    Expr* fromExpr,
+    DiagnosticSink* sink,
+    ConversionCost* outCost)
+{
+    // basic restrictions
+    if (auto someTypeDecl = isDeclRefTypeOf<SomeTypeDecl>(toType))
+    {
+        // if `some T = some U` `some T = unbound_some U`
+        if (site != CoercionSite::Argument && isDeclRefTypeOf<SomeTypeDecl>(fromType))
+        {
+            if (!isDeclRefTypeOf<UnboundSomeTypeDecl>(toType))
+            {
+                sink->diagnose(
+                    fromExpr->loc,
+                    Diagnostics::cannotAssignSomeTypeToPotentiallyDifferentSomeType);
+                return false;
+            }
+        }
+        // not allowed to assign 'dyn' to 'unboundSomeType'
```

Comment:
The comment seems inaccurate, should be 'someType`, we don't care if it is unbound or not. assigning `dyn IFoo` to `some IFoo` is always an error.